vllm/model_executor/layers/attention/backends/flash_attn.py

```diff
@@ -2,12 +2,21 @@
 from typing import List, Optional
 
 # NOTE(woosuk): This imports flash_attn under vllm/thirdparty_files/.
-from flash_attn import flash_attn_func
+from vllm.utils import is_hip
+try:
+    from flash_attn import flash_attn_func
+except ImportError:
+    if is_hip():
+        pass
+    else:
+        raise
+
 import torch
 
 from vllm.model_executor.input_metadata import InputMetadata
 from vllm.model_executor.layers.attention.ops.paged_attn import (
     PagedAttentionImpl)
+from vllm.model_executor.layers.attention.ops.flash_attention_triton import attention
```

Comment:
I have some questions:  can flash-attn-func and triton fa co-exist? or should we allow user to install them at the same time in docker file?
Also, what is the steps to validate this PR ?

Reply:
> I have some questions: can flash-attn-func and triton fa co-exist? or should we allow user to install them at the same time in docker file?

Some background

flash_attn_func is Tri Dao's "reference" [implementation](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/flash_attn_interface.py#L771) of Flash Attention. It calls the cutlass [version](https://github.com/pytorch/pytorch/blob/1a57c187602acf19e92fd1e51447b319144774a9/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L497) in the backend.

If `is_hip()`, flash_attn_func calls the CK version in the backend.

Cutlass version I believe is a pip package and one can just `pip install flash-attention`. I think for CK we need to install from our fork but I'm a little less familiar with that.

Triton is completely separate and does not intersect with either the front end flash_attn_func or cutlass / CK. It requires a Triton install. And the kernel, which is available in this PR.

So, yes both can co-exist but only one is needed.